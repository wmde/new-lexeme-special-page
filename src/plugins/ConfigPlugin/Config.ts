import { SearchedItemOption } from '@/data-access/ItemSearcher';
import {
	inject,
	InjectionKey,
} from 'vue';

export interface Config {
	readonly isAnonymous: boolean;
	readonly licenseUrl: string;
	readonly licenseName: string;
	readonly wikibaseLexemeTermLanguages: Map<string, string>;
	readonly lexicalCategorySuggestions: SearchedItemOption[];
	readonly placeholderExampleData: {
		languageLabel: string;
		lexicalCategoryLabel: string;
		lemma: string;
		spellingVariant: string;
	};
	readonly maxLemmaLength: number;
	readonly availableSearchProfiles: string[];
}

export const ConfigKey: InjectionKey<Config> = Symbol( 'Config' );

export function useConfig(): Config {
	return inject( ConfigKey, () => {
		throw new Error( 'No Config provided!' );
	}, true );
}
