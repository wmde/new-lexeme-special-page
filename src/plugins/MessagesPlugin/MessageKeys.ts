type MessageKeys
 = 'wikibaselexeme-newlexeme-lemma'
 | 'wikibaselexeme-newlexeme-lemma-empty-error'
 | 'wikibaselexeme-newlexeme-lemma-placeholder-with-example'
 | 'wikibaselexeme-newlexeme-language'
 | 'wikibaselexeme-newlexeme-language-empty-error'
 | 'wikibaselexeme-newlexeme-language-invalid-error'
 | 'wikibaselexeme-newlexeme-language-placeholder-with-example'
 | 'wikibaselexeme-newlexeme-lemma-language'
 | 'wikibaselexeme-newlexeme-lemma-language-empty-error'
 | 'wikibaselexeme-newlexeme-lemma-language-invalid-error'
 | 'wikibaselexeme-newlexeme-lemma-language-placeholder-with-example'
 | 'wikibaselexeme-newlexeme-lemma-language-help-link-target'
 | 'wikibaselexeme-newlexeme-lemma-language-help-link-text'
 | 'wikibaselexeme-newlexeme-lexicalcategory'
 | 'wikibaselexeme-newlexeme-lexicalcategory-empty-error'
 | 'wikibaselexeme-newlexeme-lexicalcategory-invalid-error'
 | 'wikibaselexeme-newlexeme-lexicalcategory-placeholder-with-example'
 | 'wikibaselexeme-newlexeme-search-existing'
 | 'wikibaselexeme-newlexeme-submit'
 | 'wikibaselexeme-newlexeme-submit-error'
 | 'wikibaselexeme-newlexeme-submitting'
 | 'wikibaselexeme-newlexeme-lemma-too-long-error'
 | 'wikibaselexeme-newlexeme-invalid-language-code-warning'
 | 'wikibase-anonymouseditwarning'
 | 'wikibase-lexeme-lemma-language-option'
 | 'wikibase-entityselector-notfound'
 | 'wikibase-shortcopyrightwarning'
 | 'copyrightpage'
 | 'wikibaselexeme-form-field-required';

export default MessageKeys;
