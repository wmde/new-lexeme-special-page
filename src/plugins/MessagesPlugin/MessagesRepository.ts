export default interface MessagesRepository {
	/**
	 * @return HTML
	 * Any bad tags in the message source are escaped.
	 */
	get( messageKey: string, ...params: readonly ( string| HTMLElement )[] ): string;

	/**
	 * @return plain text
	 * HTML-significant characters in the message source are not escaped,
	 * and the result must not be used as HTML.
	 */
	getText( messageKey: string, ...params: readonly string[] ): string;
}
