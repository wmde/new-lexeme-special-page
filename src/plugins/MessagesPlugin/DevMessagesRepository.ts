import MessageKeys from './MessageKeys';
import MessagesRepository from './MessagesRepository';

const messages: Record<MessageKeys, string> = {
	'wikibaselexeme-newlexeme-lemma': 'Lemma',
	'wikibaselexeme-newlexeme-lemma-empty-error': 'Lemma field is empty; please make an entry.',
	'wikibaselexeme-newlexeme-lemma-placeholder-with-example': 'Base form of a word, e.g. \'$1\'',
	'wikibaselexeme-newlexeme-language': 'Lexeme\'s language',
	'wikibaselexeme-newlexeme-language-empty-error': 'The Lexeme\'s language field is empty; please make an entry.',
	'wikibaselexeme-newlexeme-language-invalid-error': 'Lexeme language must be a valid Item.',
	'wikibaselexeme-newlexeme-language-placeholder-with-example': 'The Lexeme\'s language, e.g. \'$1\'',
	'wikibaselexeme-newlexeme-lexicalcategory': 'Lexical category',
	'wikibaselexeme-newlexeme-lexicalcategory-empty-error': 'The Lexical Category field is empty; please make an entry.',
	'wikibaselexeme-newlexeme-lexicalcategory-invalid-error': 'Lexical category must be a valid Item.',
	'wikibaselexeme-newlexeme-lexicalcategory-placeholder-with-example': 'The Lexeme\'s category, e.g. \'$1\'',
	'wikibaselexeme-newlexeme-lemma-language': 'Spelling variant of the Lemma',
	'wikibaselexeme-newlexeme-lemma-language-empty-error': 'The spelling variant field is empty; please make an entry.',
	'wikibaselexeme-newlexeme-lemma-language-invalid-error': 'Spelling variant must be a valid language code.',
	'wikibaselexeme-newlexeme-lemma-language-placeholder-with-example': 'Language code for the Lemma\'s spelling variant, e.g. \'$1\'',
	'wikibaselexeme-newlexeme-lemma-language-help-link-target': 'https://www.wikidata.org/wiki/Special:MyLanguage/Wikidata:Lexicographical_data/Spelling_variant',
	'wikibaselexeme-newlexeme-lemma-language-help-link-text': '(Help)',
	'wikibaselexeme-newlexeme-search-existing': 'You can check whether a Lexeme already exists by using <a href="$1">the search</a>. You can also learn more about Lexemes in the help box below.',
	'wikibaselexeme-newlexeme-submit': 'Create Lexeme',
	'wikibaselexeme-newlexeme-submit-error': 'The server encountered a temporary error and could not complete your request. Please try again.',
	'wikibaselexeme-newlexeme-submitting': 'Creating Lexeme...',
	'wikibaselexeme-newlexeme-lemma-too-long-error': 'The Lemma field is too long; please make an entry no longer than $1 characters.',
	'wikibaselexeme-newlexeme-invalid-language-code-warning': 'This Item has an unrecognized language code. Please select one below.',
	'wikibase-anonymouseditwarning': 'Warning: You are not logged in. Your IP address will be publicly visible ... If you <strong>log in</strong> or <strong>create an account</strong> ...',
	'wikibase-lexeme-lemma-language-option': '$1 ($2)',
	'wikibase-entityselector-notfound': 'No match was found',
	'wikibase-shortcopyrightwarning': 'By clicking "$1", you agree to the <a href="./$2">terms of use</a>, and you irrevocably agree to release your contribution under the <a href="$3">$4</a>.',
	copyrightpage: 'Project:Copyrights',
	'wikibaselexeme-form-field-required': 'This field is required',
};

/** Messages repository for the dev entry point. */
export default class DevMessagesRepository implements MessagesRepository {

	public get( key: MessageKeys, ...params: string[] ): string {
		return messages[ key ] !== undefined ? this.replace( messages[ key ], ...params.map( this.escape ) ) : `⧼${key}⧽`;
	}
	public getText( key: MessageKeys, ...params: string[] ): string {
		return messages[ key ] !== undefined ? this.replace( messages[ key ], ...params ) : `⧼${key}⧽`;
	}

	private replace( message: string, ...args: string[] ) {
		return message.replace(
			/\$(\d+)/g,
			( _, i ) => args[ i - 1 ],
		);
	}

	private escape( s: string ) {
		return s.replace(
			/[^0-9A-Za-z,.: ]/g,
			( c ) => '&#' + c.charCodeAt( 0 ) + ';',
		);
	}
}
