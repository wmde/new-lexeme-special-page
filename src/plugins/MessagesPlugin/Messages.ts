import MessagesRepository from '@/plugins/MessagesPlugin/MessagesRepository';
import MessageKeys from '@/plugins/MessagesPlugin/MessageKeys';
import { inject, InjectionKey } from 'vue';

/**
 * Usage:
 *
 * const messages = useMessages();
 *
 * messages.get( 'wikibaselexeme-newlexeme-lemma' )
 * messages.getUnescaped( 'wikibaselexeme-newlexeme-lemma' )
 */
export default class Messages {

	private readonly messagesRepository: MessagesRepository;

	public constructor( messagesRepository?: MessagesRepository ) {
		if ( !messagesRepository ) {
			this.messagesRepository = {
				get( messageKey ) {
					return `⧼${messageKey}⧽`;
				},
				getText( messageKey ) {
					return `⧼${messageKey}⧽`;
				},
			};
			return;
		}
		this.messagesRepository = messagesRepository;
	}

	/**
	 * Parse the message and return it as HTML, with unsafe markup escaped as necessary.
	 * Use this with v-html.
	 *
	 * @param {string} messageKey One of the well-known {@link MessageKeys}.
	 * @param {...string|HTMLElement} params Parameters for the message
	 * ($1, $2, ... in the message source).
	 * @return {string} HTML
	 */
	public get( messageKey: MessageKeys, ...params: readonly ( string|HTMLElement )[] ): string {
		return this.messagesRepository.get( messageKey, ...params );
	}

	/**
	 * Get the message, with parameters replaced but without parsing or escaping anything.
	 * Use this where the result will be escaped already, e.g. in {{}}.
	 *
	 * @param {string} messageKey One of the well-known {@link MessageKeys}.
	 * @param {...string} params Parameters for the message ($1, $2, ... in the message source).
	 * @return {string} plain text
	 */
	public getUnescaped( messageKey: MessageKeys, ...params: readonly string[] ): string {
		return this.messagesRepository.getText( messageKey, ...params );
	}
}

export const MessagesKey: InjectionKey<Messages> = Symbol( 'Messages' );

export function useMessages(): Messages {
	return inject( MessagesKey, new Messages() );
}
