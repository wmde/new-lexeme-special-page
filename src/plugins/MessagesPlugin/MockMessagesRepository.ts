import MessageKeys from './MessageKeys';
import MessagesRepository from './MessagesRepository';

const messages: Record<MessageKeys, string> = {
	'wikibaselexeme-newlexeme-lemma': 'Lemma',
	'wikibaselexeme-newlexeme-lemma-placeholder': 'Short form of a word such as in a dictionary entry, e.g. \'first\'',
	'wikibaselexeme-newlexeme-language': 'Lexeme\'s language',
	'wikibaselexeme-newlexeme-language-placeholder': 'The Lexeme\'s language, e.g. \'English\'',
	'wikibaselexeme-newlexeme-lexicalcategory': 'Lexical category',
	'wikibaselexeme-newlexeme-lexicalcategory-placeholder': 'The Lexeme\'s category, e.g. \'numeral\'',
	'wikibaselexeme-newlexeme-submit': 'Create Lexeme',
	'wikibaselexeme-newlexeme-error-no-lemma': 'Lemma field is empty, please make an entry.',
	'wikibaselexeme-newlexeme-error-lemma-is-too-long': 'FIXME (copy is missing!)',
};

export class MockMessagesRepository implements MessagesRepository {

	public get( key: MessageKeys ): string {
		return messages[ key ] !== undefined ? messages[ key ] : `⧼${key}⧽`;
	}
	public getText( key: MessageKeys ): string {
		return messages[ key ] !== undefined ? messages[ key ] : `⧼${key}⧽`;
	}
}
