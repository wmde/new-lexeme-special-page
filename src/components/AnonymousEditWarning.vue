<script setup lang="ts">
import WarningMessage from '@/components/WarningMessage.vue';
import { useConfig } from '@/plugins/ConfigPlugin/Config';
import { useMessages } from '@/plugins/MessagesPlugin/Messages';
import { computed } from 'vue';

const messages = useMessages();
const warning = computed(
	() => messages.get( 'wikibase-anonymouseditwarning' ) );
const config = useConfig();

</script>

<template>
	<warning-message
		v-if="config.isAnonymous"
		class="wbl-snl-anonymous-edit-warning"
	>
		<!-- eslint-disable-next-line vue/no-v-html -->
		<span v-html="warning" />
	</warning-message>
</template>

<style lang="scss" scoped>
.wbl-snl-anonymous-edit-warning {
	margin: 1rem 0;
}
</style>
