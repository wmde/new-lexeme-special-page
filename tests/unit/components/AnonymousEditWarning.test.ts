import AnonymousEditWarning from '@/components/AnonymousEditWarning.vue';
import { ConfigKey } from '@/plugins/ConfigPlugin/Config';
import { MessagesKey } from '@/plugins/MessagesPlugin/Messages';
import { mount } from '@vue/test-utils';

describe( 'AnonymousEditWarning', () => {

	it( 'does nothing if not anonymous', () => {
		const messages = { get: jest.fn() };
		const warning = mount( AnonymousEditWarning, {
			global: {
				provide: {
					[ ConfigKey as symbol ]: { isAnonymous: false },
					[ MessagesKey as symbol ]: messages,
				},
			},
		} );

		expect( warning.isVisible() ).toBe( false );
		expect( messages.get ).not.toHaveBeenCalled();
	} );

	it( 'matches snapshot if anonymous', () => {
		const messages = { get: jest.fn().mockImplementation( ( key ) => key ) };
		const warning = mount( AnonymousEditWarning, {
			global: {
				provide: {
					[ ConfigKey as symbol ]: { isAnonymous: true },
					[ MessagesKey as symbol ]: messages,
				},
			},
		} );

		expect( warning.html() ).toMatchSnapshot();
		expect( messages.get ).toHaveBeenCalledWith( 'wikibase-anonymouseditwarning' );
	} );

} );
