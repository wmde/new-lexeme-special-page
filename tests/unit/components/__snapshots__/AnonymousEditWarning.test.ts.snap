// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnonymousEditWarning matches snapshot if anonymous 1`] = `"<div class="wikit wikit-Message wikit-Message--warning wbl-snl-anonymous-edit-warning"><span class="wikit-Message__content"><span class="wikit wikit-Icon wikit-Icon--large wikit-Icon--warning wikit-Message__icon"><svg class="wikit-Icon__svg" xmlns="http://www.w3.org/2000/svg" fill="none" aria-hidden="true" focusable="false" viewBox="0 0 16 16"><path fill="currentColor" d="M9.163 1.68234C9.06078 1.4381 8.89901 1.22746 8.69449 1.07231C8.48997 0.917151 8.25017 0.823144 7.99999 0.800049C7.75116 0.82453 7.51294 0.919178 7.30987 1.07425C7.10679 1.22933 6.94619 1.43922 6.84459 1.68234L0.672272 13.0631C0.0337565 14.2368 0.558251 15.2 1.82768 15.2H14.1723C15.4417 15.2 15.9662 14.2368 15.3277 13.0631L9.163 1.68234ZM8.76013 12.7717H7.23986V11.1528H8.76013V12.7717ZM8.76013 9.53394H7.23986V4.67728H8.76013V9.53394Z"></path></svg></span><span><!-- eslint-disable-next-line vue/no-v-html --><span>wikibase-anonymouseditwarning</span></span></span></div>"`;
